### ISSUE TITLE
FIXME: Fixing unfair write behavior of SocketChannelIOHelper#batch

### ISSUE THREAD

**Issue by haruntuncay** [TooTallNate/Java-WebSocket#873]
Current behavior of SocketChannelIOHelper#batch writes all the queued out buffers at one go, which may cause other "write ready" channels to wait if the total size of queued buffers exceeds socket's internal output buffer.

The behavior of` WritableByteChannel#write` is commented out like so: (which can be found [here](https://docs.oracle.com/javase/7/docs/api/java/nio/channels/WritableByteChannel.html#write(java.nio.ByteBuffer)))

> Unless otherwise specified, a write operation will return only after writing all of the r requested bytes. Some types of channels, depending upon their state, may write only some of the bytes or possibly none at all. A socket channel in non-blocking mode, for example, cannot write any more bytes than are free in the socket's output buffer.

Currently working on this, and open to suggestions 😄 .
I believe a better behaviour should try to write the buffer, and if the write operation returns 0, indicating that the socket's internal buffer is full, batch method should return and let other write ready channels to continue.

**Comment by haruntuncay** [comment]
Hi @marci4,
I don't know how I didn't realize before but



	/*int written = */sockchannel.write( buffer );
	if( buffer.remaining() > 0 ) {
		return false;
	}
This part already does what I had as a solution in my mind. 
What kind of solution do you think would work better than this ? 
And should I close the issue for now since that my solution is already covered by this behavior ?

**Comment by marci4** [comment]
Does this fix the problem the write will return only after everything is written?

**Comment by haruntuncay** [comment]
I believe so, yes. This behavior should return early with false, if the socket's internal output buffer is filled and should not hold control any longer.

**Comment by marci4** [comment]
Sounds ok for me to close this then as cannot reproduce :)

Thank you!

Best regards,
Marcel
